test/unit/event_machine_proxy_test.rb

```diff
@@ -78,18 +78,41 @@ def synchrony(&block)
   end
 
   describe "#defer" do
-    it "should output its block's output when it doesn't raise an error" do
-      ExceptionHandling.logger = Logger.new(STDERR)
+    before do
+      logger = Logger.new(STDERR)
+      logger.extend ContextualLogger::LoggerMixin
+      ExceptionHandling.logger = logger
+    end
 
+    it "should output its block's output when it doesn't raise an error, by default" do
       @em.run do
         assert_equal 12, @em.defer("#defer success") { 12 }
         @em.stop
       end
     end
 
-    it "should raise an error when its block raises an error" do
-      ExceptionHandling.logger = Logger.new(STDERR)
+    it "should not wait for its block to run if option is passed" do
+      @block_ran = false
+
+      @em.run do
+        assert_nil @em.defer("#defer success", wait_for_result: false) { @block_ran = true; 12 }
+        assert_equal false, @block_ran
+        @em.add_periodic_timer(0.1) { @em.stop if @em.defers_finished? }
+      end
+
+      assert_equal true, @block_ran
+    end
+
+    it "should handle exceptions when not waiting for its block to run" do
+      mock(ExceptionHandling).log_error(is_a(RuntimeError), "defer", {})
 
+      @em.run do
+        assert_nil @em.defer("#defer success", wait_for_result: false) { raise RuntimeError, "error in defer" }
+        @em.add_periodic_timer(0.1) { @em.stop if @em.defers_finished? }
```

Comment:
Interesting way to do it! If this test ever fails, it will just hang, though. Bertand Meyer would recommend a counter:
```
check_finished_counter = 0
@em.add_periodic_timer(0.1) do
  (check_finished_counter += 1) > 20 and raise "defer never finished!"
  @em.defers_finished? and @em.stop
end
```

Reply:
I stole the idea from EventMachine's unit test as I was curious how they were testing the non-blocking defer. I've added the counter version as a helper method, to ensure the tests don't hang